app/services/fastly_vcl/whitelisted_params.rb

```diff
@@ -0,0 +1,67 @@
+module FastlyVCL
+  # Handles updates to our VCL snippet on Fastly that whitelists params
+  class WhitelistedParams
+    VCL_DELIMITER_START = "^(".freeze
+    VCL_DELIMITER_END = ")$".freeze
+    SNIPPET_NAME = "Whitelist certain querystring parameters".freeze
+    FILE_PARAMS = YAML.load_file("config/fastly/whitelisted_params.yml").freeze
+
+    class << self
+      def update
+        fastly = Fastly.new(api_key: ApplicationConfig["FASTLY_API_KEY"])
+        service = fastly.get_service(ApplicationConfig["FASTLY_SERVICE_ID"])
+        latest_version = service.version
+
+        snippet = fastly.get_snippet(ApplicationConfig["FASTLY_SERVICE_ID"],
+                                     latest_version.number,
+                                     SNIPPET_NAME)
+
+        current_params = params_to_array(snippet.content)
+
+        return unless params_updated?(current_params)
+
+        new_version = latest_version.clone
+        new_snippet = fastly.get_snippet(ApplicationConfig["FASTLY_SERVICE_ID"],
+                                         new_version.number,
+                                         SNIPPET_NAME)
+        new_snippet.content = build_content(FILE_PARAMS, new_snippet.content)
+        new_snippet.save!
+
+        new_version.activate!
+        log_params_diff_to_datadaog(current_params, new_version)
+        Rails.logger.info("Fastly updated to version #{new_version.number}.")
+      end
+
+      private
+
+      def params_updated?(current_params)
+        (current_params - FILE_PARAMS).any? || (FILE_PARAMS - current_params).any?
+      end
+
+      def params_to_array(snippet_content)
+        snippet_suffix = snippet_content.split(VCL_DELIMITER_START).last
+        fastly_params = snippet_suffix.split(VCL_DELIMITER_END).first
+
+        fastly_params.split("|")
+      end
+
+      def build_content(new_params, snippet_content)
+        new_params = new_params.join("|")
+        snippet_prefix = snippet_content.split(VCL_DELIMITER_START).first
+        snippet_suffix = snippet_content.split(VCL_DELIMITER_END).last
+
+        "#{snippet_prefix}#{VCL_DELIMITER_START}#{new_params}#{VCL_DELIMITER_END}#{snippet_suffix}"
```

Comment:
So this is just the Regex snippet, does Fastly then just magically know to use this with `regfilter_except` function while leaving the rest of the snippet like the `if` URL statement alone?

Reply:
It's actually rebuilding the entire VCL here. I did that on purpose so all other updates are ignored and if we wanted to make changes to VCL syntax around the params, those would be done by an admin directly in Fastly and everything would flow through as normal.

In other words, the finished product here is:
```
import querystring;
sub vcl_recv {
    # return this URL with only the parameters that match this regular expression
    if (req.url !~ "/internal/" && req.url !~ "/search/") {
      set req.url = querystring.regfilter_except(req.url, "^(a_id|args|article_id|...)$");
    }
}
```

And that looks like this:
```
snippet prefix
=> "import querystring;\nsub vcl_recv {\n    # return this URL with only the parameters that match this regular expression\n    if (req.url !~ \"/internal/\" && req.url !~ \"/search/\") {\n      set req.url = querystring.regfilter_except(req.url, \""

VCL_DELIMITER_START
=> "^("

new_params
=> "a_id|args|article_id|..."

VCL_DELIMITER_END
=> ")$"

snippet_suffix
=> "\");\n    }\n}"
```

Kind of hard to read in this format, but we are just re-using whatever we get from Fastly except for the param list - that's the only piece in the VCL we update with this PR.

I considered moving the VCL snippet into the code as a template. But then we have 2 versions of the VCL code to maintain - the template in code and what we have in Fastly.

Reply:
OH I get it now, thanks for the explanation!